src/streams.c

```diff
@@ -81,67 +83,121 @@ static Int READ_COMMAND(Obj *evalResult)
  */
-Obj FuncREAD_ALL_COMMANDS( Obj self, Obj stream, Obj echo )
+Obj FuncREAD_ALL_COMMANDS(Obj self, Obj instream, Obj echo, Obj outputFunc)
 {
     ExecStatus status;
-    Int resultCount, resultCapacity;
-    UInt dualSemicolon;
-    Obj result, resultList;
-    Obj evalResult;
+    Int        resultCount, resultCapacity;
+    UInt       dualSemicolon;
+    Obj        result, resultList;
+    Obj        evalResult;
+    Obj        outstream;
+    Obj        outstreamString;
+    Obj        outstreamStringCopy;
+    Obj        outputFunc_return;
+
+    /* try to open the streams */
+    if (!OpenInputStream(instream, echo == True)) {
+        return Fail;
+    }
 
-    /* try to open the stream */
-    if (!OpenInputStream(stream, echo == True)) {
+
+    if (outputFunc != False) {
+        outstreamString = NEW_STRING(0);
+        outstream = DoOperation2Args(ValGVar(GVarName("OutputTextString")),
+                                     outstreamString, True);
+    }
+    else {
+        outstreamString = False;
+        outstream = False;
+    }
+    if (outstream != False && !OpenOutputStream(outstream)) {
+        CloseInput();
         return Fail;
     }
 
     resultCount = 0;
     resultCapacity = 16;
-    resultList = NEW_PLIST( T_PLIST, resultCapacity );
+    resultList = NEW_PLIST(T_PLIST, resultCapacity);
     SET_LEN_PLIST(resultList, resultCount);
 
     do {
         ClearError();
-        status = ReadEvalCommand(STATE(BottomLVars), &evalResult, &dualSemicolon);
 
-        if(!(status & (STATUS_EOF | STATUS_QUIT | STATUS_QQUIT))) {
+        if (outputFunc != False) {
+            // Clean in case Callback function had output
+            SET_LEN_STRING(outstreamString, 0);
+        }
+
+        status =
+            ReadEvalCommand(STATE(BottomLVars), &evalResult, &dualSemicolon);
+
+        if (!(status & (STATUS_EOF | STATUS_QUIT | STATUS_QQUIT))) {
             resultCount++;
             if (resultCount > resultCapacity) {
                 resultCapacity += 16;
                 GROW_PLIST(resultList, resultCapacity);
             }
-            result = NEW_PLIST( T_PLIST, 3 );
-            SET_LEN_PLIST(result, 1);
+            if (outputFunc != False) {
+                outstreamStringCopy = CopyToStringRep(outstreamString);
+                SET_LEN_STRING(outstreamString, 0);
+                result = NEW_PLIST(T_PLIST, 5);
+                SET_LEN_PLIST(result, 5);
+                SET_ELM_PLIST(result, 5, outstreamStringCopy);
+                CHANGED_BAG(result);
+            }
+            else {
+                result = NEW_PLIST(T_PLIST, 3);
+                SET_LEN_PLIST(result, 3);
+            }
             SET_ELM_PLIST(result, 1, False);
-            SET_ELM_PLIST(resultList, resultCount, result );
-            SET_LEN_PLIST(resultList, resultCount );
+            SET_ELM_PLIST(resultList, resultCount, result);
+            CHANGED_BAG(result_list);
+            SET_LEN_PLIST(resultList, resultCount);
```

Comment:
... and then replace the preceding three lines by `PushPlist(resultList, result);` (which grows the list, adds the new entry, adjusts the list length and even calls `CHANGED_BAG` for you.